### PR TITLE
88 Completed auth flow

### PR DESCRIPTION
**Pull request by ishootdice**
- SignIn and SignUp components rework
- Moved styles from SignIn and SignUp to AuthPageComponent styles 
- Created endpoint for user update on backend side, modified UserService
- Made forms inside SignIn and SignUp templates more responsive
- Changed routing after SignIn and SignUp
- Created couple entities on Backend for user update

### CODE REVIEW

**Comment by ishootdice** [comment]
Screenshot 2023-03-28 at 17 18 41

**Comment by ishootdice** [comment]
Screenshot 2023-03-28 at 17 19 40

**Review by NadiaMaksymchuk — APPROVED** [review]

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs [hunk 1]
```diff
@@ -5,32 +5,50 @@
...
+    public async Task<UserDto> UpdateCurrentUser(UserDto userDto)
+    {
+        var user = _mapper.Map<User>(userDto);
+        user.Id = _authService.UserId;
 
-            return _mapper.Map<UserDto>(user);
+        if (userDto.Tags != null)
+        {
+            var tagsNames = userDto.Tags.Select(x => x.Name).ToList();
+            user.Tags = _context.Tags.Where(t => tagsNames.Contains(t.Name)).ToList();
```
Comment (tatianahutii) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:46]: await

File: frontend/src/app/modules/auth/sign-in/sign-in.component.ts [hunk 2]
```diff
@@ -24,50 +21,48 @@ export class SignInComponent {
...
-        email: new FormControl(this.email),
-        password: new FormControl(this.password),
+        email: new FormControl('', [Validators.required, Validators.email, Validators.maxLength(30)]),
+        password: new FormControl(
+            '',
+            [
+                Validators.required,
+                Validators.minLength(6),
+                Validators.maxLength(25),
+                Validators.pattern('^(?=.*[a-z])(?=.*[A-Z])[a-zA-Z\\d@$!%*?&\\.]{6,}$'),
```
Comment (tatianahutii) [inline frontend/src/app/modules/auth/sign-in/sign-in.component.ts:31]: please create a separate shared file for regex storing

File: frontend/src/app/modules/auth/sign-up/sign-up.component.ts [hunk 3]
```diff
@@ -1,181 +1,148 @@
...
+        private userService: UserService,
+        private router: Router,
     ) {
         super();
-        this.Countries = this.countriesTzLangProvider.getCountriesList();
+        this.countries = countriesTzLangProvider.getCountriesList().map((x) => x.name);
+        this.languages = countriesTzLangProvider.getLanguagesList();
+        this.englishLevels = Object.values(EnglishLevel) as string[];
+        this.sexOptions = Object.values(this.sexEnumeration) as string[];
+        this.ages = Ages;
```
Comment (tatianahutii) [inline frontend/src/app/modules/auth/sign-up/sign-up.component.ts:65]: move to a separate method

File: frontend/src/app/modules/auth/sign-up/sign-up.component.ts [hunk 4]
```diff
@@ -1,181 +1,148 @@
...
-        }
-
-        return '';
+    private signUp() {
+        this.authService
+            .signUp(this.email.value, this.password.value)
+            .then(() => {
+                this.createUser();
+                this.toastr.success('Successfully sign up', 'Sign up');
+                this.router.navigate(['topics']);
```
Comment (tatianahutii) [inline frontend/src/app/modules/auth/sign-up/sign-up.component.ts:80]: show a notification after an user creating

File: frontend/src/app/modules/user-profile/select-topics-page/select-topics-page.component.ts [hunk 5]
```diff
@@ -1,21 +1,51 @@
...
+        private userService: UserService,
+        private toastr: ToastrService,
+        private router: Router,
+    ) {
+        super();
+        this.userService.getUser()
+            .pipe(this.untilThis)
+            .subscribe((resp) => {
+                this.currentUser = resp;
+            });
```
Comment (tatianahutii) [inline frontend/src/app/modules/user-profile/select-topics-page/select-topics-page.component.ts:32]: to a separate method

**Review by tatianahutii — CHANGES_REQUESTED** [review]
Looks good. Only a few small fixes

**Comment by Anatolii-Ivanov** [comment]
From screenshot
Only selected field should have this blue border/shadow
Can you make right part the same way it's done on landing page, please

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs [hunk 6]
```diff
@@ -5,32 +5,50 @@
...
+
+    public async Task<UserDto> UpdateCurrentUser(UserDto userDto)
+    {
+        var user = _mapper.Map<User>(userDto);
+        user.Id = _authService.UserId;
 
-            return _mapper.Map<UserDto>(user);
+        if (userDto.Tags != null)
+        {
+            var tagsNames = userDto.Tags.Select(x => x.Name).ToList();
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:45]: Maybe we can just assign tags to existing user without changing other fields?

File: frontend/src/app/modules/auth/sign-in/sign-in.component.html [hunk 7]
```diff
@@ -1,28 +1,90 @@
...
+                        <img src="/assets/user-profile/alert-triangle.svg" alt="">
+                        Field is required
+                    </mat-error>
+        
+                    <mat-error class="error" *ngIf="this.email.errors?.['email'] 
+                    && (this.email.dirty || this.userLoginForm.submitted)">
+                        <img src="/assets/user-profile/alert-triangle.svg" alt="">
+                        Wrong format
+                    </mat-error>
+                </div>
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/sign-in/sign-in.component.html:25]: Can you make it via code and then just show correct exception here?

File: frontend/src/app/modules/auth/sign-in/sign-in.component.html [hunk 8]
```diff
@@ -1,28 +1,90 @@
...
+                        Minimal length is 6 characters
+                    </mat-error>
+
+                    <mat-error class="error" *ngIf="this.password.errors?.['maxlength'] 
+                    && (this.password.dirty || this.userLoginForm.submitted)">
+                        <img src="/assets/user-profile/alert-triangle.svg" alt="">
+                        Maximal length is 25 characters
+                    </mat-error>
+                </div>
+
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/sign-in/sign-in.component.html:70]: The same

File: frontend/src/app/modules/auth/sign-up/sign-up.component.html [hunk 9]
```diff
@@ -1,81 +1,287 @@
...
+                    <img src="/assets/user-profile/alert-triangle.svg" alt="">
+                    Field is required
+                </mat-error>
+    
+                <mat-error class="error" *ngIf="this.email.errors?.['email'] 
+                && (this.email.dirty || this.userRegisterForm.submitted)">
+                    <img src="/assets/user-profile/alert-triangle.svg" alt="">
+                    Wrong format
+                </mat-error>
+            </div>
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/sign-up/sign-up.component.html:25]: The same for all errors here (You can create two functions, one to check if it should be shown, second - what should be shown

File: frontend/src/app/modules/auth/sign-up/sign-up.component.ts (hunk 3 above)
Comment (Anatolii-Ivanov) [reply frontend/src/app/modules/auth/sign-up/sign-up.component.ts:65]: and ngOnInit

**Comment by ishootdice** [comment]
Screenshot 2023-03-29 at 03 29 25

**Review by NadiaMaksymchuk — APPROVED** [review]

File: backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Controllers/UserController.cs [hunk 10]
```diff
@@ -34,13 +35,20 @@ public async Task<ActionResult<UserDto>> Post([FromBody] UserRegisterDto user)
...
         }
 
-
         [HttpPut("{userId}")]
         public ActionResult<UserDto> Update(int userId, UserDto userDto)
         {
             return Ok(userDto);
         }
 
+        [HttpPost]
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Controllers/UserController.cs:44]: Please, don't create endpoint with same url even if they have different parameters
You can name this one /users/tags

**Review by Anatolii-Ivanov — DISMISSED** [review]

File: backend/EasySpeak.Core/EasySpeak.Core.DAL/Entities/User.cs [hunk 11]
```diff
@@ -22,15 +22,14 @@ public class User : Entity<long>
         public ICollection<Chat> Chats { get; private set; }
         public ICollection<Lesson> Lessons { get; private set; }
         public ICollection<Lesson> CreatedLessons { get; private set; }
-        public ICollection<Tag> Tags { get; private set; }
+        public ICollection<Tag> Tags { get; set; } = new List<Tag>();
```
Comment (tatianahutii) [inline backend/EasySpeak.Core/EasySpeak.Core.DAL/Entities/User.cs:25]: please fix warnings in this file

File: frontend/src/app/modules/auth/sign-up/sign-up.component.ts [hunk 12]
```diff
@@ -1,181 +1,173 @@
...
+            .pipe(this.untilThis)
+            .subscribe(() => {
+                this.toastr.success('Account successfully registered', 'Succes!');
+            });
+
+        this.authService
+            .signUp(this.email.value, this.password.value)
+            .then(() => {
+                this.toastr.success('Successfully sign up', 'Sign up');
+                this.router.navigate(['topics']);
```
Comment (tatianahutii) [inline frontend/src/app/modules/auth/sign-up/sign-up.component.ts:97]: this won't work as you expected

File: frontend/src/app/modules/user-profile/select-topics-page/select-topics-page.component.ts (hunk 5 above)
Comment (tatianahutii) [reply frontend/src/app/modules/user-profile/select-topics-page/select-topics-page.component.ts:36]: still, need this

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by NadiaMaksymchuk — DISMISSED** [review]

File: frontend/src/app/modules/auth/sign-up/sign-up.component.ts [hunk 13]
```diff
@@ -1,181 +1,174 @@
...
+            .subscribe(
+                () => {
+                    this.toastr.success('Account successfully registered', 'Succes!');
+                    this.toastr.success('Successfully sign up', 'Sign up');
+                    this.router.navigate(['topics']);
+                },
+                (error) => {
+                    this.toastr.error(error.message, 'Sign up');
+                },
+            );
```
Comment (tatianahutii) [inline frontend/src/app/modules/auth/sign-up/sign-up.component.ts:102]:
```suggestion
     this.authService.signUp(this.email.value, this.password.value)
         .then(() => {
             this.createUser()
                  .pipe( this.untilThis,)
                  .subscribe(
                      () => {
                          this.toastr.success('Account successfully registered', 'Succes!');
                           this.router.navigate(['topics']);
                       },
                       (error) => {
                           this.toastr.error(error.message, 'Sign up');
                       })
                   })

```

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Pull request merged**